pkg/store/prometheus.go

```diff
@@ -434,25 +429,26 @@ func (p *PrometheusStore) startPromRemoteRead(ctx context.Context, q *prompb.Que
 	return presp, nil
 }
 
-// matchesExternalLabels filters out external labels matching from matcher if exists as the local storage does not have them.
-// It also returns false if given matchers are not matching external labels.
-func matchesExternalLabels(ms []storepb.LabelMatcher, externalLabels labels.Labels) (bool, []storepb.LabelMatcher, error) {
-	if len(externalLabels) == 0 {
-		return true, ms, nil
-	}
-
-	tms, err := storepb.TranslateFromPromMatchers(ms...)
+// matchesExternalLabels returns false if given matchers are not matching external labels.
+// If true, matchesExternalLabels also returns Prometheus matchers without those matching external labels.
+func matchesExternalLabels(ms []storepb.LabelMatcher, externalLabels labels.Labels) (bool, []*labels.Matcher, error) {
+	tms, err := storepb.MatchersToPromMatchers(ms...)
 	if err != nil {
 		return false, nil, err
 	}
 
-	var newMatcher []storepb.LabelMatcher
+	if len(externalLabels) == 0 {
+		return true, tms, nil
+	}
```

Comment:
IMO this length check should be before line 435? We don't need to allocate convert matchers in this case.

Reply:
We might need to, as we want to return translated slice.

Reply:
You are right. I missed that.